### PR TITLE
feat(context): compact oversized workflow context (#229)

### LINKED ISSUE #229: Automated Context Window Management
## What problem does this solve?

Amelia has no strategy for managing context window limits during long workflows. As conversations grow:

1. ~~Token usage increases linearly with conversation length~~ ✅ Token usage is tracked per-agent with cache-aware cost calculation
2. No compaction or summarization of old context
3. Eventually hits model context limits
4. ~~No visibility into context utilization~~ ✅ Dashboard shows aggregate cost/token trends and per-workflow breakdowns (but not context window fill level)

## Who would benefit from this?

- Users running complex multi-phase workflows
- Teams working on large codebases with extensive tool output
- Anyone hitting context limit errors

## Current state

The following building blocks already exist:

- **Token usage tracking** (`amelia/server/models/tokens.py`): `TokenUsage` and `TokenSummary` models track input/output/cache tokens per agent, with cache-aware cost calculation for 15+ models
- **Usage analytics API** (`amelia/server/routes/usage.py`): `GET /api/usage` with date filtering, period-over-period comparison, per-model breakdown
- **Dashboard cost visualization** (`dashboard/src/pages/CostsPage.tsx`, `dashboard/src/components/UsageCard.tsx`): trend charts, model sparklines, per-workflow token tables, CSV export
- **Driver usage extraction** (`amelia/drivers/base.py`): `DriverUsage` model with `get_usage()` on driver interface
- **Tool output truncation** (`amelia/drivers/api/deepagents.py`): 100KB byte-based truncation with `truncated` flag
- **Task-based context scoping** (`amelia/pipelines/implementation/nodes.py`): multi-task plans clear `driver_session_id` between tasks, reducing per-iteration context
- **Brainstorm token eviction** (`amelia/server/services/brainstorm.py`): `tool_token_limit_before_evict=20000` — only existing token budget mechanism, limited to brainstorming

## What does success look like?

1. **Compaction strategies**: Configurable approaches to reduce context size
   - Keep last N turns (simplest)
   - Token-based truncation (count backward from limit)
   - Recursive summarization (LLM-generated summaries replace old content)
2. **Async compaction**: Background processing to avoid blocking user
3. **Compaction tracking**: Know which events are summarized vs original
4. **Token budget**: Per-workflow and per-phase token limits with alerts

### Phase 1: Context Window Awareness

Token counting and cost tracking already exist. Remaining work:

- [ ] Add model context window size metadata alongside existing pricing data (e.g. Claude Sonnet = 200K, GPT-4o = 128K)
- [ ] Track real-time context utilization during agent execution (how full is the window)
- [ ] Dashboard context window gauge/bar showing fill level per active workflow
- [ ] Alerts when approaching model context limits (configurable threshold)

### Phase 2: Simple Compaction

- [ ] Configurable "keep last N turns" strategy for `ImplementationState.tool_calls`/`tool_results` (currently append-only via `operator.add` reducers with no pruning)
- [ ] Token-aware tool output truncation (current 100KB byte-based truncation in `ApiDriver` is not token-aware and only applies to the API driver)
- [ ] Automatic compaction triggering based on token threshold
- [ ] Token budget configuration — per-profile or per-agent context policies (e.g. `AMELIA_MAX_CONTEXT_TOKENS`, profile-level settings)

### Phase 3: Smart Compaction

- [ ] LLM-generated summaries of old conversation segments
- [ ] Preserve key decisions and context (priority classification of conversation content)
- [ ] Track provenance (summary vs original) — events have `EventLevel` but no "summarized" marker
- [ ] Async/background compaction without blocking agent execution

## Constraints and considerations

- Compaction must preserve critical context (user decisions, key findings)
- Should be transparent to user (show what was compacted)
- Different strategies for different content types (tool output vs reasoning)
- `ApiDriver._sessions` stores full conversation in `MemorySaver` with no truncation — driver-level integration needed
- CLI driver delegates context management to the external tool entirely

## Key integration points

| Component | File | Why it matters |
|-----------|------|----------------|
| ImplementationState | `amelia/pipelines/implementation/state.py` | Append-only `tool_calls`/`tool_results` need pruning support |
| Developer._build_prompt() | `amelia/agents/developer.py` | Where conversation is formatted — needs context budget awareness |
| ApiDriver._sessions | `amelia/drivers/api/deepagents.py` | In-memory `MemorySaver` stores full history with no truncation |
| DriverInterface | `amelia/drivers/base.py` | `execute_agentic()` needs context budget parameter |
| EventBus | `amelia/server/events/bus.py` | Existing retention pattern could extend to compaction tracking |
| Model pricing | `amelia/server/models/tokens.py` | Already has 15+ models — add context window sizes here |

## References

Based on enterprise agentic systems white papers covering:
- Context engineering as core architecture
- Session history management patterns
- Token bloat mitigation strategies

### PR DESCRIPTION
**Pull request by anderskev**
Problem:
Amelia could measure context usage after #505 but still had no strategy to compact long workflow history before model limits were hit.

Change:
- Adds simple context compaction for implementation tool_calls/tool_results with protected first turn, recent turns, and explicit omitted-turn markers.
- Replaces append-only reducers with a controlled ReplaceList path for compaction rewrites.
- Adds token-aware sandbox output truncation.
- Triggers Developer compaction when context utilization crosses the configured threshold and emits context_compacted events.
- Keeps LLM-generated summaries deferred.

Verification:
- uv run ruff check --fix amelia tests
- uv run mypy amelia
- unset SSL_CERT_FILE; GIT_CONFIG_GLOBAL=/dev/null uv run pytest tests/unit/
- push pre-push hook: ruff, mypy, 2623 passed, dashboard build passed

Depends on #661
Closes #229

### CODE REVIEW

**Comment by anderskev** [comment]
@coderabbitai review this pr

File: amelia/agents/developer.py [hunk 1]
```diff
@@ -188,6 +195,48 @@ async def run(
...
+                        "AMELIA_CONTEXT_COMPACTION_THRESHOLD",
+                        str(DEFAULT_COMPACTION_THRESHOLD),
+                    )
+                )
+                keep_last = int(
+                    os.environ.get(
+                        "AMELIA_CONTEXT_KEEP_LAST_TURNS",
+                        str(DEFAULT_KEEP_LAST_N_TURNS),
+                    )
+                )
```
[the remark this answers is not in the document]
Comment (anderskev) [reply amelia/agents/developer.py]: Fixed in `9800ddf1`. Env var parsing for `AMELIA_CONTEXT_COMPACTION_THRESHOLD` and `AMELIA_CONTEXT_KEEP_LAST_TURNS` is now done eagerly at the top of `Developer.run()` via `_parse_compaction_threshold()` and `_parse_keep_last_turns()`, before streaming starts. Non-numeric values raise `ValueError` with a clear message; threshold is range-checked to (0, 1] and keep_last to >= 1. This prevents a crash after the success event was already yielded.

File: amelia/drivers/api/deepagents.py [hunk 2]
```diff
@@ -44,14 +45,41 @@
 from amelia.logging import log_claude_result, log_todos
 
 
-__all__ = ["ApiDriver", "LocalSandbox"]
+__all__ = ["ApiDriver", "LocalSandbox", "truncate_text_to_token_budget"]
 
 # Maximum output size before truncation (100KB)
 _MAX_OUTPUT_SIZE = 100_000
+# Default token budget for shell/tool output returned to the model.
+_MAX_OUTPUT_TOKENS = int(os.environ.get("AMELIA_MAX_TOOL_OUTPUT_TOKENS", "12000"))
```
[the remark this answers is not in the document]
Comment (anderskev) [reply amelia/drivers/api/deepagents.py]: Fixed in `9800ddf1`. The import-time `int()` call is now wrapped in `_parse_max_output_tokens()`, which raises a clear `ValueError` ("AMELIA_MAX_TOOL_OUTPUT_TOKENS must be a non-negative integer, got {raw!r}") on non-numeric input and rejects negative values. The driver module no longer crashes with a generic int parse error.

File: amelia/drivers/api/deepagents.py [hunk 3]
```diff
@@ -44,14 +45,41 @@
...
+    enc = tokenizer or _default_tokenizer()
+    tokens = enc.encode(text)
+    original_tokens = len(tokens)
+    if original_tokens <= max_tokens:
+        return text, False, original_tokens
+
+    marker = f"\n... [tool output truncated to {max_tokens} tokens from {original_tokens}]"
+    marker_tokens = enc.encode(marker)
+    content_budget = max(0, max_tokens - len(marker_tokens))
+    truncated = enc.decode(tokens[:content_budget]) + marker
```
[the remark this answers is not in the document]
Comment (anderskev) [reply amelia/drivers/api/deepagents.py:111]: Fixed in `9800ddf1`. `truncate_text_to_token_budget` now checks if the marker itself (encoded) exceeds `max_tokens`. If it does, the function returns just the first `max_tokens` tokens of content with no marker appended, guaranteeing the returned string never exceeds `max_tokens` when encoded. Added tests for the tiny-budget and zero-budget edge cases.

File: amelia/drivers/api/deepagents.py [hunk 4]
```diff
@@ -68,6 +96,17 @@ class LocalSandbox(FilesystemBackend, SandboxBackendProtocol):
...
+    def __init__(
+        self,
+        *args: Any,
+        max_output_tokens: int | None = None,
+        tokenizer: tiktoken.Encoding | None = None,
+        **kwargs: Any,
+    ) -> None:
+        super().__init__(*args, **kwargs)
+        self.max_output_tokens = max_output_tokens or _MAX_OUTPUT_TOKENS
+        self.tokenizer = tokenizer or _default_tokenizer()
```
[the remark this answers is not in the document]
Comment (anderskev) [reply amelia/drivers/api/deepagents.py:147]: Fixed in `9800ddf1`. `LocalSandbox.__init__` now distinguishes `None` (use default `_MAX_OUTPUT_TOKENS`) from invalid values: if `max_output_tokens <= 0` it raises `ValueError` ("max_output_tokens must be a positive integer"). Explicit `0` no longer silently falls back to the default.

File: amelia/pipelines/implementation/context_compaction.py [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+    reason_text = f" Reason: {reason}." if reason else ""
+    return (
+        "[CONTEXT COMPACTED] "
+        f"Omitted {len(omitted_calls)} middle tool turn(s) to keep the workflow within "
+        f"the model context window.{reason_text} "
+        f"Protected first turn and the most recent turns are preserved. "
+        f"Omitted range: {first}..{last}. "
+        f"Tool names seen: {names or 'unknown'}. "
+        f"Approx omitted result words: {result_tokens}."
+    )
```
[the remark this answers is not in the document]
Comment (anderskev) [reply amelia/pipelines/implementation/context_compaction.py:97]: Fixed in `9800ddf1`. Added `_aggregate_prior_markers()` which scans `omitted_calls` for prior compaction markers: it sums their `omitted_turns` metadata into the new total, and uses the first/last real (non-marker) call IDs as `first_omitted_call_id` / `last_omitted_call_id`. Marker calls are skipped when computing first/last omitted IDs. The event data, marker `tool_input`, and marker output string all use the aggregated values. Added tests for re-compaction and the aggregation helper.

**Comment by anderskev** [comment]
## Closing without merge — redundant with driver-layer compaction

Closing this PR (and its tracking issue #229) because the model-context compaction it aims to provide already exists at the driver layer:

- **ApiDriver:** deepagents' `SummarizationMiddleware` is hardcoded into `create_deep_agent` and fires at ~85% of the model context window by default.
- **CLI drivers:** the external Claude Code / Codex CLI manages its own context.

The core issue with this change: it compacts `ImplementationState.tool_calls`/`tool_results`, which are **Amelia bookkeeping that is never sent to the model** — `Developer._build_prompt()` doesn't read them, and the real conversation lives in deepagents' `MemorySaver` (ApiDriver) or the external tool (CLI). So the new reducer + `CONTEXT_COMPACTED` event shrink an internal object and emit a dashboard event but reduce **zero** model-facing tokens, while running at the same 0.85 threshold deepagents already acts on. CLI drivers were spared only because their `get_usage()` returns `context_utilization=None` — an accidental guard, not a designed one.

The one non-redundant piece — token-aware tool-output truncation in `LocalSandbox` — is marginal (deepagents summarization already compacts oversized output, and a 100KB byte cap already exists), not worth a new reducer, state-schema change, and event type.

If we later hit a real single-turn context blowout, we'll do a narrow truncation fix then. See #229 for the full rationale.
